File: slack-reader/src/main/java/dev/number6/slackreader/adaptor/SnsNotificationAdaptor.java
package dev.number6.slackreader.adaptor;

import dev.number6.slackreader.port.SlackReaderConfigurationPort;
import dev.number6.slackreader.port.NotificationPort;
import dev.number6.message.ChannelMessages;
import com.amazonaws.services.sns.AmazonSNS;
import com.amazonaws.services.sns.model.PublishRequest;
import com.google.gson.Gson;

public class SnsNotificationAdaptor implements NotificationPort {

    private final Gson gson = new Gson();
    private final AmazonSNS sns;
    private final String topicArn;

    public SnsNotificationAdaptor(SlackReaderConfigurationPort config, AmazonSNS sns) {
        this.topicArn = config.getTopicArn();
        this.sns = sns;
    }

    @Override
    public void broadcast(ChannelMessages messages) {
        sns.publish(new PublishRequest(topicArn, gson.toJson(messages)));
    }
}

File: slack-reader/src/main/kotlin/dev/number6/slackreader/adaptor/SnsNotificationAdaptor.kt
package dev.number6.slackreader.adaptor

import com.amazonaws.services.sns.AmazonSNS
import com.amazonaws.services.sns.model.PublishRequest
import com.google.gson.Gson
import dev.number6.message.ChannelMessages
import dev.number6.slackreader.port.NotificationPort
import dev.number6.slackreader.port.SlackReaderConfigurationPort

class SnsNotificationAdaptor(config: SlackReaderConfigurationPort, private val sns: AmazonSNS) : NotificationPort {
    private val gson: Gson = Gson()
    private val topicArn: String = config.getTopicArn()

    override fun broadcast(messages: ChannelMessages) {
        sns.publish(PublishRequest(topicArn, gson.toJson(messages)))
    }

}